cmd/networkagent/main.go

```diff
@@ -0,0 +1,453 @@
+		NicID:       nicID,
+	}
+	if _, err := s.client.DeleteNICWithVF(ctx, deleteReq); err != nil {
+		// best effort
+		log.G(ctx).WithField("nicID", nicID).Warn("failed to delete endpoint nic")
+	}
+
+	// then remove the VF from the UVM
+	removeVFReq := &ncproxygrpc.RemoveVFRequest{
+		ContainerID:          req.ContainerID,
+		DeviceID:             nicID,
+		VirtualFunctionIndex: s.conf.NetworkingSettings.NICWithVFSettings.VirtualFunctionIndex,
+	}
+	if _, err := s.client.RemoveVF(ctx, removeVFReq); err != nil {
+		return nil, err
+	}
+
+	delete(s.containerToNICVirtualFunctionID, req.ContainerID)
+	return &nodenetsvc.ConfigureNetworkingResponse{}, nil
+}
+
+func (s *service) teardownHNSHelper(ctx context.Context, req *nodenetsvc.ConfigureNetworkingRequest, containerNamespaceID string) (*nodenetsvc.ConfigureNetworkingResponse, error) {
+	eReq := &ncproxygrpc.GetEndpointsRequest{}
+	resp, err := s.client.GetEndpoints(ctx, eReq)
+	if err != nil {
+		return nil, err
+	}
+	for _, endpoint := range resp.Endpoints {
+		if endpoint.Namespace == containerNamespaceID {
+			nicID, ok := s.endpointToNicID[endpoint.Name]
+			if !ok {
+				log.G(ctx).WithField("name", endpoint.Name).Warn("endpoint was not assigned a NIC ID previously")
+				continue
+			}
+			// remove endpoints that are in the namespace as NICs
+			nsReq := &ncproxygrpc.DeleteNICRequest{
+				ContainerID:  req.ContainerID,
+				NicID:        nicID,
+				EndpointName: endpoint.Name,
+			}
+			if _, err := s.client.DeleteNIC(ctx, nsReq); err != nil {
+				log.G(ctx).WithField("name", endpoint.Name).Warn("failed to delete endpoint nic")
+			}
+			delete(s.endpointToNicID, endpoint.Name)
+		}
+	}
+	return &nodenetsvc.ConfigureNetworkingResponse{}, nil
+}
+
+func (s *service) modifyHelper(ctx context.Context, containerID, containerNamespaceID string, iovSettings *ncproxygrpc.IovEndpointPolicySetting) error {
+	eReq := &ncproxygrpc.GetEndpointsRequest{}
+	resp, err := s.client.GetEndpoints(ctx, eReq)
+	if err != nil {
+		return err
+	}
+	for _, endpoint := range resp.Endpoints {
+		if endpoint.Namespace == containerNamespaceID {
+			nicID, ok := s.endpointToNicID[endpoint.Name]
+			if !ok {
+				return fmt.Errorf("endpoint was not assigned a NIC ID previously")
+			}
+
+			req := &ncproxygrpc.ModifyNICRequest{
+				ContainerID:       containerID,
+				NicID:             nicID,
+				EndpointName:      endpoint.Name,
+				IovPolicySettings: iovSettings,
+			}
+			if _, err := s.client.ModifyNIC(ctx, req); err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+func (s *service) ConfigureNetworking(ctx context.Context, req *nodenetsvc.ConfigureNetworkingRequest) (*nodenetsvc.ConfigureNetworkingResponse, error) {
+	containerNamespaceID, ok := s.containerToNamespace[req.ContainerID]
+	if !ok {
+		return nil, fmt.Errorf("no namespace was previously created for containerID %s", req.ContainerID)
+	}
+
+	log.G(ctx).WithField("req", req).Info("ConfigureNetworking request")
```

Comment:
I'd move this to the top so we know from logs we hit this function even if there's no namespace